orchestra/static/orchestra/todos/todo-list.directive.es6.js

```diff
@@ -98,9 +99,26 @@ export default function todoList (orchestraApi) {
         todoApi.delete(todo)
       }
 
+      todoList.addActionToTodoActivityLog = (todo, action) => {
+        const datetimeUtc = moment.tz(moment(), moment.tz.guess()).utc()
+        var activityLog = JSON.parse(todo.activity_log.replace(/'/g, '"'))
+        activityLog['actions'].push({
+          'action': action,
+          'datetime': datetimeUtc.format('YYYY-MM-DD HH:mm:ss')
+        })
+        todo.activity_log = JSON.stringify(activityLog).replace(/'/g, '"')
+      }
+
+      todoList.onToggleTodo = (todo, collapsed) => {
+        todoList.addActionToTodoActivityLog(todo, collapsed ? 'collapse' : 'expand')
+        todoApi.update(todo)
```

Comment:
given the messages happen asynchronously, it's possible that activity log [a,b,c] arrives before [a,b] and we lose c.

are we OK with that? if not, do we want to add logic on the backend to only log appends to the activity log? i'm personally OK losing a few activities---just wanted to raise the idea that we might :)

Reply:
Technically yes that can happen. But on the interface ideally, it should happen since the user can only take one action at a time. For example, an item cannot be toggled and skipped at the same time. I would say the time lag to update activity log should be very low. 

I am OK with that.